File: CPP/ByTier/D/2075_D_Equalization.cpp
// Problem: CF 2075 D - Equalization
// https://codeforces.com/contest/2075/problem/D

/*
D. Equalization
Algorithms/Techniques: Dynamic Programming with Bit Manipulation, Preprocessing
Time Complexity: O(N^2 * M + t) where N = 57 (number of bits), M = 15 (max k value), t = number of test cases
Space Complexity: O(N^2) for DP table

The problem involves making two numbers x and y equal by dividing them with powers of 2,
where each operation costs 2^k and each k can only be used once.

We precompute the minimum cost to make any pair of bit representations (x >> i, y >> j) equal.
This approach uses dynamic programming on bits, considering all possible operations (dividing x or y by 2^k).
*/

#include <stdio.h>
#include <string.h>
#include <algorithm>

using namespace std;
typedef long long LL;
const int N = 57, M = 15; // N is number of bits for values up to 10^17, M is max k value
const int inf = 0x3f3f3f3f;
int t;
int dp[N + 1][N + 1]; // dp[i][j] stores min cost to equalize x>>i and y>>j
LL x, y;

void solve() {
  scanf("%lld%lld", &x, &y);
  int ans = inf;
  // Try all bit representations of x and y to find minimum cost to make them equal
  for (int i = 0; i <= N; i++)
    for (int j = 0; j <= N; j++)
      if ((x >> i) == (y >> j)) // If the bit-shifted versions are equal
        ans = min(ans, dp[i][j]); // Update answer with minimum cost
  printf("%d\n", ans);
}

int main() {
  memset(dp, 0x3f, sizeof(dp)); // Initialize dp with infinity
  dp[0][0] = 0; // Base case: no cost to equalize x>>0 and y>>0 (both are original values)
  
  // Preprocessing step: compute minimum cost to bring any two bit shifts to same value
  for (int i = 1; i <= M; i++) // For each possible k from 1 to 15
    for (int j = N; j >= 0; j--) // Traverse backwards to avoid recomputation
      for (int k = N; k >= 0; k--) {
        if (j >= i)
          dp[j][k] = min(dp[j][k], dp[j - i][k] + (1 << i)); // Update if dividing x reduces j
        if (k >= i)
          dp[j][k] = min(dp[j][k], dp[j][k - i] + (1 << i)); // Update if dividing y reduces k
      }
  
  scanf("%d", &t); // Read number of test cases
  while (t--)
    solve(); // Process each test case
  return 0;
}


// https://github.com/VaHiX/codeForces/

File: Python/ByTier/D/2075_D_Equalization.py
# Problem: CF 2075 D - Equalization
# https://codeforces.com/contest/2075/problem/D

"""
D. Equalization

Problem Description:
Given two non-negative integers x and y, we can perform operations to make them equal.
Each operation allows choosing a positive integer k and dividing either x or y by 2^k (rounded down),
with the cost of 2^k. Each value of k may be used at most once.

Algorithm:
- Uses dynamic programming with bit manipulation.
- Precomputes minimum cost states using a function f(u,v) = u << 7 ^ v to map 2D to 1D.
- The DP tracks minimal cost for reducing x and y down to various bit lengths.
- Optimizes by considering only bits that differ between x and y.

Time Complexity: O(30^2 * 2^14) ~ O(10^5)
Space Complexity: O(2^14) ~ O(16384)

"""

import io
import os
import sys

input = io.BytesIO(os.read(0, os.fstat(0).st_size)).readline


def f(u, v):
    return u << 7 ^ v  # Bit manipulation to compress (u,v) into a single index


t = int(input())
ans = []
inf = 1 << 62
l = 30
dp = [inf] * (1 << 14)
dp[0] = 0  # Base case: no operations needed when both x and y are 0
for i in range(1, l):  # For each bit position k from 1 to 29
    p = 1 << i  # Cost of operation using bit k (2^k)
    for j in range(3 * l, -1, -1):  # Traverse all valid combinations of (j,k) in reverse order
        for k in range(3 * l, -1, -1):
            mi = inf
            if i <= j:  # Check if we can use this bit on x side
                mi = min(mi, dp[f(j - i, k)] ^ p)  # Use operation on x
            if i <= k:  # Check if we can use this bit on y side
                mi = min(mi, dp[f(j, k - i)] ^ p)  # Use operation on y
            dp[f(j, k)] = min(mi, dp[f(j, k)])  # Update DP state

mi = list(dp)  # Copy DP values to use later for final lookup
for i in range(3 * l, -1, -1):  # Propagate minimal costs forward to avoid overcounting
    for j in range(3 * l, -1, -1):
        mi[f(i, j)] = min(mi[f(i, j)], mi[f(i + 1, j)], mi[f(i, j + 1)])

for _ in range(t):
    x, y = map(int, input().split())
    n, m = x.bit_length(), y.bit_length()  # Get number of bits needed to represent x and y
    ans0 = mi[f(n, m)]  # Minimum cost to reduce both to same bit lengths
    
    # While leading bits match, continue reducing until they differ
    while min(n, m) and x >> (n - 1) == y >> (m - 1):
        n, m = n - 1, m - 1  # Reduce bit-lengths
        ans0 = min(ans0, dp[f(n, m)])  # Update with new minimal cost
    
    ans.append(ans0)
sys.stdout.write("\n".join(map(str, ans)))


# https://github.com/VaHiX/codeForces/